### PR TITLE
Optimize redundant logic used in queries

### PR DESCRIPTION
**Pull request by pdiaz**
After upgrading parse-server from 2.7.4 to 4.4.0, MongoDB 4.0 maxed CPU usage from what used to be barely 20% usage.

This does reduce the redundant complexity on queries that involve pointer permissions.

### CODE REVIEW

**Comment by dplewis** [comment]
Thanks for the PR? Can you write a test case for this?

**Comment by mtrezza** [comment]
Would you please also open an issue for this to explain the proposed change? This allow others to join the discussion more easily and helps us to better trace changes in Parse Server.

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 ### master
 [Full Changelog](https://github.com/parse-community/parse-server/compare/4.4.0...master)
+- IMPROVE: Optimize queries on classes with pointer permissions.
```
Comment (mtrezza) [inline CHANGELOG.md:5]: Add: `[#7061](https://github.com/parse-community/parse-server/pull/7061). Thanks to [Pedro Diaz](https://github.com/pdiaz)`

You can remove your name of course if you prefer to do so.

File: src/Controllers/DatabaseController.js [hunk 2]
```diff
@@ -1365,6 +1365,32 @@ class DatabaseController {
...
+        );
+        if (foundEntries === queries[shorter].length) {
+          // If the shorter query is completely contained on the longer one, we can skip it.
+          query.$or.splice(longer, 1);
+          queries.splice(longer, 1);
+        }
+      }
+    }
+
+    if (query.$or.length === 1) return query.$or[0];
```
Comment (mtrezza) [inline src/Controllers/DatabaseController.js:1389]: Can you please wrap the if block in brackets?

@dplewis Shouldn't lint currently enforce this? If not, I think we should add that rule; if blocks without brackets are just an unnecessary bug risk IMHO.

File: src/Controllers/DatabaseController.js [hunk 3]
```diff
@@ -1365,6 +1365,32 @@ class DatabaseController {
...
+
```
Comment (mtrezza) [inline src/Controllers/DatabaseController.js:1390]: Please remove empty line.

File: src/Controllers/DatabaseController.js [hunk 4]
```diff
@@ -1448,13 +1474,23 @@ class DatabaseController {
...
+          const queryClauseEntries = this.objectToEntriesStrings(queryClause);
+          const queryEntries = this.objectToEntriesStrings(query);
+          const foundEntries = queryEntries.reduce(
+            (acc, entry) => acc + (queryClauseEntries.includes(entry) ? 1 : 0),
+            0
+          );
+          if (foundEntries === queryClauseEntries.length) {
+            return query;
+          }
+
```
Comment (mtrezza) [inline src/Controllers/DatabaseController.js:1486]: Please remove empty line.

**Review by mtrezza — CHANGES_REQUESTED** [review]
Some minor comments

File: src/Controllers/DatabaseController.js [hunk 5]
```diff
@@ -1365,6 +1365,65 @@ class DatabaseController {
...
+  // This helps to create intermediate objects for simpler comparison of
+  // key value pairs used in query objects. Each key value pair will represented
+  // in a similar way to json
+  objectToEntriesStrings(query: any): Array<string> {
+    return Object.entries(query).map(a => a.map(s => JSON.stringify(s)).join(':'));
+  }
+
+  reduceOrOperation(query: { $or: Array<any> }): any {
+    if (!query.$or) {
+      return query;
```
Comment (dplewis) [inline src/Controllers/DatabaseController.js:1377]: Can you add write test case for if `$or` doesn't exist?

File: src/Controllers/DatabaseController.js [hunk 6]
```diff
@@ -1365,6 +1365,65 @@ class DatabaseController {
...
+    }
+    if (query.$or.length === 1) {
+      return query.$or[0];
+    }
+    return query;
+  }
+
+  reduceAndOperation(query: { $and: Array<any> }): any {
+    if (!query.$and) {
+      return query;
```
Comment (dplewis) [inline src/Controllers/DatabaseController.js:1403]: Can you add write test case for if `$and` doesn't exist?

**Review by dplewis — CHANGES_REQUESTED** [review]
Just a few nits

**Comment by dplewis** [comment]
Have you tried your branch in production?

**Comment by mtrezza** [comment]
Before merging, mind any ongoing discussion in https://github.com/parse-community/parse-server/issues/7065.

**Comment by pdiaz** [comment]
> Have you tried your branch in production or tested it against your large database?

Yes, after this optimization performance went back to the same range as before.

**Review by dplewis — APPROVED** [review]
LGTM! @mtrezza How does it look?

**Comment by mtrezza** [comment]
I see that the test cases test the new methods of the database controller, but there is no high-level test case that tests whether these methods are implemented correctly within the controller. Can we add a test case that makes a Parse.Query and evaluates the optimized query that would be sent to the DB client?

**Comment by pdiaz** [comment]
> I see that the test cases test the new reduction methods of the database controller, but there is no high-level test case that tests whether these methods are implemented correctly within the controller and in which scenarios they are actually applied to a query. Can we add some test cases that make a `ParseQuery` and ensure the optimization of the resulting query?

Good call! I previously tested only with two pointer permissions, so I added two extra test cases for 2 and 3 pointer permissions on a class and fixed and issue detected with the second case.

**Review by mtrezza — CHANGES_REQUESTED** [review]
Just trivia, but on a general note:

Does this optimization apply to all $or / $and across a query or only related pointer permissions?

File: spec/DatabaseController.spec.js [hunk 7]
```diff
@@ -236,6 +236,65 @@ describe('DatabaseController', function () {
...
+        .and.returnValue({ type: 'Object' });
+
+      const output = databaseController.addPointerPermissions(
+        schemaController,
+        CLASS_NAME,
+        OPERATION,
+        query,
+        ACL_GROUP
+      );
+
```
[the remark this answers is not in the document]
Comment (pdiaz) [reply spec/DatabaseController.spec.js:292]: All empty lines? If not between what exactly?

**Comment by pdiaz** [comment]
> Does this optimization apply to all $or / $and across a query or only related pointer permissions?

@mtrezza No all, only on find operations related to pointer permissions and not the queries themselves. Update operations could also have unoptimized $and operations.

A full optimizer for queries in general would require many more test cases.

**Comment by mtrezza** [comment]
> A full optimizer for queries in general would require many more test cases.

I was thinking into that direction. A Parse Query Optimizer would be an interesting feature, but maybe that's more a job for the guys at MongoDB.

This PR looks good to me, ready for merge I'd say.

**Comment by pdiaz** [comment]
> > A full optimizer for queries in general would require many more test cases.
> 
> I was thinking into that direction. A Parse Query Optimizer would be an interesting feature, but maybe that's more a job for the guys at MongoDB.
> 
> This PR looks good to me, ready for merge I'd say.

Leaving that task to MongoDB guys means that a MongoDB upgrade would be required for everybody and that would take even much longer. I've been already working on that idea in my own, it will take some time to be ready. 

This PR right now is already a good step forward.

**Comment by pdiaz** [comment]
@mtrezza @dplewis Looks like PG tests are a bit flaky. Can you trigger a restart?

**Comment by mtrezza** [comment]
Checks passed. @dplewis any thoughts before merging?

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/5.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0](https://github.com/parse-community/parse-server/releases/tag/5.0.0)

**Pull request merged**